File: LeetCode/121.kt
// USER: SIDHARTH MUDGIL
// DATE: Tuesday 12, Sep 2023
// TASK: 121. Best Time to Buy and Sell Stock

class Solution {
    fun maxProfit(prices: IntArray): Int {
       var max = 0
       var min = prices[0]

       for (i in 1 ..< prices.size) {
           max = maxOf(max, prices[i] - min)
           min = minOf(min, prices[i])
       }

       return max
    }
}

File: leetcode/121.cpp
// USER: SIDHARTH MUDGIL
// DATE: Friday 20, Jan 2023
// TASK: 121. Best Time to Buy and Sell Stock

#include <iostream>
#include <vector>
#include <string>

using namespace std;

int maxProfit(vector<int> &prices)
{
    int min = INT_MAX;
    int res = 0;
    ;

    for (int i = 0; i < prices.size(); i++)
    {
        if (prices[i] < min)
            min = prices[i];

        res = max(res, prices[i] - min);
    }

    return res;
}

int main()
{
    vector<int> p{1, 22, 52, 6, 34, 3};
    cout << maxProfit(p);
    return 0;
}